objects/src/notes/inputs.rs

```diff
@@ -0,0 +1,60 @@
+use super::{Digest, Felt, Hasher, NoteError, ZERO};
+
+/// Holds the inputs which are placed onto the stack before a note's script is executed.
+/// - inputs are stored in reverse order such that when they are pushed onto stack they are in
+///   the correct order
+/// - hash is computed from inputs in stack order
+#[derive(Debug, Eq, PartialEq)]
+pub struct NoteInputs {
+    inputs: [Felt; 16],
+    hash: Digest,
+}
+
+impl NoteInputs {
+    /// Number of note inputs.
+    const NOTE_NUM_INPUTS: usize = 16;
+
+    // CONSTRUCTOR
+    // --------------------------------------------------------------------------------------------
+    /// Returns NoteInputs created from the provided inputs.
+    ///
+    /// The inputs are padded with ZERO such that they are always of length 16.
+    ///
+    /// # Errors
+    /// Returns an error if the number of provided inputs is greater than 16.
+    pub fn new(inputs: &[Felt]) -> Self {
+        if inputs.len() > Self::NOTE_NUM_INPUTS {
+            NoteError::too_many_inputs(inputs.len());
+        }
+
+        // pad inputs with ZERO to be constant size (16 elements)
+        let mut padded_inputs = [ZERO; Self::NOTE_NUM_INPUTS];
+        for (i, input) in inputs.iter().enumerate() {
+            padded_inputs[i] = *input;
+        }
+
+        // compute hash from padded inputs in stack order.
+        let hash = Hasher::hash_elements(&padded_inputs);
+
+        // inputs stored in reverse order such that they are in correct order when pushed onto stack.
+        padded_inputs.reverse();
```

Comment:
I wonder if we should hash after reversing. I'm imagining these inputs will be written into kernel memory using `adv_pipe` instruction, and then before the note is executed, we'll read the inputs from memory onto the stack using `mem_loadw` instruction.

So, let's say our inputs are `(a, b)` and we want the stack for the note script to look like `[b, a, ...]`. This means that we probably want the inputs stored in memory like so:
```
[i + 0]: [0, 0, 0, 0]
[i + 1]: [0, 0, 0, 0]
[i + 2]: [0, 0, 0, 0]
[i + 3]: [0, 0, a, b]
```
With this arrangement, we should be able to use 4 `mem_loadw` instructions (as well as some `padw` and other instructions) to get the desired stack state.

This means that the hash we'd want to compute would be of:
```
[0, 0, 0, 0,  0, 0, 0, 0,  0, 0, 0, 0,  0, 0, a, b]
```
But double-check my logic here as I didn't think it through completely.